ddtrace/internal/coverage/threading_coverage.py

```diff
@@ -0,0 +1,70 @@
+"""Provides functionality for context-based coverage to work across threads
+
+Without this, context-based collection in the parent process would not capture code executed by threads (due to the
+parent process' context variables not being shared in threads).
+
+The collection of coverage is done when the thread's join() method is called, so context-level coverage will not be
+captured if join() is not called.
+
+Since the ModuleCodeCollector is already installed at the process level, there is no need to reinstall it or ensure that
+its include_paths are set.
+
+Session-level coverage does not need special-casing since the ModuleCodeCollector behavior is process-wide and
+thread-safe.
+"""
+from queue import Queue
+import threading
+
+from ddtrace.internal.coverage.code import ModuleCodeCollector
+
+
+Thread = threading.Thread
+thread_init = Thread.__init__
+thread_boostrap_inner = Thread._bootstrap_inner  # type: ignore[attr-defined]
+thread_join = Thread.join
+
+DD_PATCH_ATTR = "_datadog_patch"
+
+
+def _is_patched():
+    return hasattr(threading, DD_PATCH_ATTR)
+
+
+class CoverageCollectingThread(threading.Thread):
+    def __init__(self, *args, **kwargs):
+        """Wraps the thread initialization creation to enable coverage collection
+
+        Only enables coverage if the parent process' context-level coverage is enabled.
+        """
+        self._should_cover = ModuleCodeCollector.is_installed() and ModuleCodeCollector.coverage_enabled_in_context()
+
+        if self._should_cover:
+            self._coverage_queue = Queue()
+
+        thread_init(self, *args, **kwargs)
+
+    def _bootstrap_inner(self):
+        """Collect thread-level coverage data in a context and queue it up for the parent process to absorb"""
+        if self._should_cover:
+            self._coverage_context = ModuleCodeCollector.CollectInContext()
+            self._coverage_context.__enter__()
+
+        thread_boostrap_inner(self)
+
+        if self._should_cover:
+            covered_lines = ModuleCodeCollector.get_context_data_json()
+            self._coverage_context.__exit__()
+            self._coverage_queue.put(covered_lines)
+
+    def join(self, *args, **kwargs):
+        """Absorb coverage data from the thread after it's joined"""
```

Comment:
Would it be more effective to send out data during the thread's life, not when it's joined? I am a bit worried that Thread could collect a lot of data and then it will take some time to serialize it to msgpack and send it out

Reply:
It could be, yes, but that would involve reorganizing the parent thread to have a consumer thread (or having a queue that accumulates everything and then gets flushed at the end, which IMO would be more or less equivalent to this current implementation).

We have plans to introduce a multiprocessing-friendly shared memory API in the not-too-distance future that would enable better behavior for multithreading/multiprocessing, so I'm holding off on moving to a producer/consumer model until then.